### PR TITLE
[UX][Serve] More comprehensive `target_qps_per_replica` check

### LINKED ISSUE #3359: [Serve] Service update scales to zero unexpectedly?
Repro:
- Change https://github.com/skypilot-org/skypilot/blob/master/examples/serve/http_server/task.yaml#L13 to `replicas: 1`
- `sky serve up -n http examples/serve/http_server/task.yaml  -y`
- In another shell, `while 1; do curl -L 34.239.227.235:30001; sleep 0.5; done`
- Wait until the replica is READY and the curl's above show responses
- Change the YAML again to 
```
  # replicas: 1
  replica_policy:
    min_replicas: 0
    max_replicas: 1
```
- `sky serve update http examples/serve/http_server/task.yaml  -y`

Observed: Status immediately became
```
Services
NAME  VERSION  UPTIME  STATUS      REPLICAS  ENDPOINT
http  -        6m 48s  NO_REPLICA  0/1       34.239.227.235:30001

Service Replicas
SERVICE_NAME  ID  VERSION  IP              LAUNCHED    RESOURCES       STATUS         REGION
http          1   1        44.213.109.175  7 mins ago  1x AWS(vCPU=2)  SHUTTING_DOWN  us-east-1
```
After a while it became 
```
Services
NAME  VERSION  UPTIME  STATUS      REPLICAS  ENDPOINT
http  -        8m 14s  NO_REPLICA  0/0       34.239.227.235:30001

Service Replicas
No existing replicas.
```
The curl's are still ongoing and fail to wake up the service.

Both problems seem unexpected. Since there's constant traffic hitting the service, scale-to-zero shouldn't have kicked in. Also, why is the service not woken up?

commit 82c50f5c281fc88e

cc @cblmemo @Michaelvll

### PR DESCRIPTION
**Pull request by cblmemo**
Fixes #3359

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
1. Autoscaling enabled
```yaml
service:
  readiness_probe:
    path: /health
    initial_delay_seconds: 20
  replica_policy:
    min_replicas: 1
    max_replicas: 2
    target_qps_per_replica: 1
```
```bash
$ sky serve up -n http @temp/svc.yaml
Service from YAML spec: @temp/svc.yaml
Service Spec:
Readiness probe method:           GET /health
Readiness initial delay seconds:  20
Replica autoscaling policy:       Autoscaling from 1 to 2 replicas with Target 1 QPS/replica
Spot Policy:                      No spot policy
```
2. Error (`target_qps_per_replica` not set)
```yaml
service:
  readiness_probe:
    path: /health
    initial_delay_seconds: 20
  replica_policy:
    min_replicas: 1
    max_replicas: 2
```
```bash
$ sky serve up -n http @temp/svc.yaml
Service from YAML spec: @temp/svc.yaml
ValueError: Detect different min_replicas and max_replicas while target_qps_per_replica is not set. To enable autoscaling, please set target_qps_per_replica.
```
3. Fixed replicas when min==max
```yaml
service:
  readiness_probe:
    path: /health
    initial_delay_seconds: 20
  replica_policy:
    min_replicas: 1
    max_replicas: 1
    target_qps_per_replica: 1
```
```bash
$ sky serve up -n http @temp/svc.yaml
Service from YAML spec: @temp/svc.yaml
Service Spec:
Readiness probe method:           GET /health
Readiness initial delay seconds:  20
Replica autoscaling policy:       Fixed 1 replica
Spot Policy:                      No spot policy
```
4. Autoscaling disabled
```bash
service:
  readiness_probe:
    path: /health
    initial_delay_seconds: 20
  replica_policy:
    min_replicas: 1
    max_replicas: 1
```
```bash
$ sky serve up -n http @temp/svc.yaml
Service from YAML spec: @temp/svc.yaml
Service Spec:
Readiness probe method:           GET /health
Readiness initial delay seconds:  20
Replica autoscaling policy:       Fixed 1 replica
Spot Policy:                      No spot policy
```
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/serve/service_spec.py [hunk 1]
```diff
@@ -37,27 +37,44 @@ def __init__(
...
-        if target_qps_per_replica is not None and max_replicas is None:
-            with ux_utils.print_exception_no_traceback():
-                raise ValueError('max_replicas must be set where '
-                                 'target_qps_per_replica is set.')
+        if target_qps_per_replica is not None:
+            if max_replicas is None:
+                with ux_utils.print_exception_no_traceback():
+                    raise ValueError('max_replicas must be set where '
+                                     'target_qps_per_replica is set.')
+            if max_replicas == min_replicas:
```
Comment (concretevitamin) [inline sky/serve/service_spec.py:50]: I feel this should be allowed (closed range). People may set min=max when testing.

File: sky/serve/service_spec.py [hunk 2]
```diff
@@ -37,27 +37,44 @@ def __init__(
...
+                        'target_qps_per_replica is set. No autoscaling will '
+                        'be performed. Please set different min_replicas and '
+                        'max_replicas to enable autoscaling.')
+        else:
+            if max_replicas is not None and max_replicas != min_replicas:
+                with ux_utils.print_exception_no_traceback():
+                    raise ValueError(
+                        'Detect different min_replicas and max_replicas '
+                        'while target_qps_per_replica is not set. To enable '
+                        'autoscaling, please set target_qps_per_replica.')
```
Comment (concretevitamin) [inline sky/serve/service_spec.py:63]:
```suggestion
                        'Detected different min_replicas and max_replicas '
                        'while target_qps_per_replica is not set. To enable '
                        'autoscaling, please set target_qps_per_replica.')
```

File: sky/serve/service_spec.py [hunk 3]
```diff
@@ -236,10 +253,13 @@ def autoscaling_policy_str(self):
...
             return f'Fixed {self.min_replicas} replica{min_plural}'
+        # Already checked in __init__.
+        assert self.target_qps_per_replica is not None
         # TODO(tian): Refactor to contain more information
         max_plural = '' if self.max_replicas == 1 else 's'
-        return (f'Autoscaling from {self.min_replicas} to '
-                f'{self.max_replicas} replica{max_plural}')
+        return (f'Autoscaling from {self.min_replicas} to {self.max_replicas} '
+                f'replica{max_plural} with Target {self.target_qps_per_replica}'
+                ' QPS/replica')
```
Comment (concretevitamin) [inline sky/serve/service_spec.py:262]:
```suggestion
        return (f'Autoscaling from {self.min_replicas} to {self.max_replicas} '
                f'replica{max_plural} (target QPS per replica: {self.target_qps_per_replica})')
```
How about this?

**Review by concretevitamin — APPROVED** [review]
Thanks @cblmemo!

File: sky/serve/service_spec.py (hunk 1 above)
Comment (cblmemo) [reply sky/serve/service_spec.py:50]: Good point! Enabled.

**Pull request merged**